### PR TITLE
#160370620 Edit Users Trip Request

### PR DESCRIPTION
**Pull request by seunzone**
#### What does this PR do?
This PR contains the implementation for updating trip requests made by a user.
Trip requests could be a one way, return or multiple trip.

#### Description of Task to be completed?
- create reducers, action & sagas to edit a user trip request
- follow design mockup for edit tables
- write tests for editing a users trip request

#### How should this be manually tested?
- Clone this repo - `git clone https://github.com/andela/travel_tool_front.git`
- Switch to the repo directory - `cd travel_tool_front`
- Check out this branch - `git checkout ft-edit-trips-request-160370620 `
- Install dependencies - `yarn install`
- Start the server  - `make start`
- Navigate to `localhost:3000` in a browser
- Log in with your Andela mail
- Navigate to `my requests` pane
- Select the dotted buttons and click edit on the request you wish to make changes to

#### Any background context you want to provide?
- A request has to be made before one can think of editing, hence you have to first create a request before you can test out this feature.
- Accepted and rejected trip requests cannot be edited
- Managers cannot edit a request

#### What are the relevant pivotal tracker stories?
N/A

#### Screenshots (if appropriate)
screen shot 2018-09-27 at 4 52 08 pm
screen shot 2018-09-27 at 4 52 32 pm
screen shot 2018-09-27 at 4 53 38 pm

#### Questions:

### CODE REVIEW

File: src/components/Forms/NewRequestForm/NewRequestForm.jsx [hunk 1]
```diff
@@ -199,7 +261,7 @@ class NewRequestForm extends PureComponent {
       const addedTripStateValues = this.getDefaultTripStateValues(parentIds);
       return {
         parentIds: parentIds + 1,
-        trips: trips.concat([{}]),
```
Comment (johnmutuma5) [inline src/components/Forms/NewRequestForm/NewRequestForm.jsx]: And this one too, @seunzone

File: src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx [hunk 2]
```diff
@@ -329,16 +356,16 @@ describe('<NewRequestForm />', () => {
...
+    expect(wrapper.instance().state.values['origin-2']).toBe('');
     expect(wrapper.instance().state.parentIds).toBe(3);
     wrapper.instance().removeTrip(1);
     expect(wrapper.instance().state.parentIds).toBe(2);
     // still have one trip, origin-1 should now be what was at origin-2 and so forth
-    expect(wrapper.instance().state.values[`origin-1`]).toBe('');
+    expect(wrapper.instance().state.values['origin-1']).toBe('');
     // after shifting state values, origin-{parentIds} in state should be undefined
-    expect(wrapper.instance().state.values[`origin-2`]).toBe(undefined);
-    expect(wrapper.instance().state.trips).toHaveLength(2);
```
Comment (johnmutuma5) [inline src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx]: @seunzone, this is the test that safeguards that stability. Let's keep it.

File: src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx [hunk 3]
```diff
@@ -329,16 +356,16 @@ describe('<NewRequestForm />', () => {
...
     wrapper.instance().addNewTrip();
     wrapper.instance().addNewTrip();
-    expect(wrapper.instance().state.values[`origin-1`]).toBe('');
-    expect(wrapper.instance().state.values[`origin-2`]).toBe('');
+    expect(wrapper.instance().state.values['origin-1']).toBe('');
+    expect(wrapper.instance().state.values['origin-2']).toBe('');
     expect(wrapper.instance().state.parentIds).toBe(3);
     wrapper.instance().removeTrip(1);
     expect(wrapper.instance().state.parentIds).toBe(2);
     // still have one trip, origin-1 should now be what was at origin-2 and so forth
```
Comment (johnmutuma5) [inline src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx]: @seunzone, I made a typo on this comment earlier, it's supposed to read:
`// still have two trips, ....`
 Please adjust it for me. Thank you

File: src/components/Forms/NewRequestForm/NewRequestForm.jsx [hunk 4]
```diff
@@ -116,18 +167,25 @@ class NewRequestForm extends PureComponent {
...
+      let parentIds, secondTripStateValues = {};
+      if(modalType === 'edit request') {
+        parentIds = requestOnEdit.trips.length;
+      } else {
+        parentIds = 2;
+        secondTripStateValues = this.getDefaultTripStateValues(1);
+      }
       this.setState(prevState => ({
-        parentIds : 2,
-        trips: [].concat([prevState.trips[0] || {}, {}]),
```
[the remark this answers is not in the document]
Comment (seunzone) [reply src/components/Forms/NewRequestForm/NewRequestForm.jsx]: Thanks for your feedback @johnmutuma5.
Kindly test out the feature by adding multiple trips, you'll see that this current implementation works and all the data are been maintained in the state and it does not break.
This also works well when you try to edit the data as well.

File: src/components/Forms/NewRequestForm/NewRequestForm.jsx (hunk 1 above)
Comment (seunzone) [reply src/components/Forms/NewRequestForm/NewRequestForm.jsx]: Thanks, @johnmutuma5 . Kindly refer to my previous comment. Concatenating this empty object adds an object of an empty trip to the array of the list of trips to be submitted to the API.
In fact, the API would return an error stating it cannot submit a trip with an empty origin, destination, etc. You can sync with me to get more clarity on the matter. Thanks.

File: src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx (hunk 3 above)
Comment (seunzone) [reply src/components/Forms/NewRequestForm/__tests__/NewRequestForm.test.jsx]: ok, will do.

**Comment by seunzone** [comment]
@johnmutuma5 Thanks for pointing out edge cases and thanks for the guidance. All set. Also, I set the handleClearForm() to return an empty object in the test. Hope this does not affect anything as I expect to clear the form of all details therein.

**Comment by seunzone** [comment]
Thanks for pointing it out @clintfidel 
I've updated it as requested.

File: src/components/Forms/NewRequestForm/NewRequestForm.jsx [hunk 5]
```diff
@@ -116,18 +166,28 @@ class NewRequestForm extends PureComponent {
...
     this.setState({
       selection: tripType,
     }, this.validate);
     if (tripType === 'multi' && !collapse) {
       this.collapsible();
-      const secondTripStateValues = this.getDefaultTripStateValues(1);
+      let parentIds, trips, secondTripStateValues = {};
+      if(modalType === 'edit request') {
+        parentIds = requestOnEdit.trips.length;
+        trips = requestOnEdit.trips;
```
Comment (johnmutuma5) [inline src/components/Forms/NewRequestForm/NewRequestForm.jsx]: @seunzone, superb. All working well now.

Comment (seunzone) [reply src/components/Forms/NewRequestForm/NewRequestForm.jsx]: Thanks @johnmutuma5

**Comment by samuelkubai** [comment]
Hi @seunzone, due to other stories being merged a few seconds ago there are conflicts of files on this PR. Please resolve conflicts so that I may review this.

**Comment by seunzone** [comment]
@samuelkubai I just fixed the merge conflicts
